crates/codegen/runtime/cargo/src/runtime/bindings/mod.rs

```diff
@@ -1,10 +1,193 @@
+
+    use crate::cst::KindTypes;
+
+    pub type File = ast::File<KindTypes>;
+    pub type Functions = functions::Functions<KindTypes>;
+    pub type Graph = graph::Graph<KindTypes>;
+
+    pub use metaslang_graph_builder::{ExecutionConfig, NoCancellation, Variables};
+}
+
+use std::collections::{BTreeSet, HashMap};
+use std::fmt;
+use std::fmt::Debug;
+use std::iter::once;
+use std::path::PathBuf;
+
+use metaslang_graph_builder::stack_graph;
+use semver::Version;
+use stack_graphs::graph::StackGraph;
+use stack_graphs::partial::PartialPaths;
+use stack_graphs::stitching::{ForwardPartialPathStitcher, GraphEdgeCandidates, StitcherConfig};
+use thiserror::Error;
 
 use crate::cst::KindTypes;
+use crate::cursor::Cursor;
+
+type Builder<'a> = stack_graph::Builder<'a, KindTypes>;
+type GraphHandle = stack_graphs::arena::Handle<stack_graphs::graph::Node>;
+
+#[derive(Error, Debug)]
+pub enum BindingsError {
+    #[error(transparent)]
+    IoError(#[from] std::io::Error),
+
+    #[error("Unknown error condition: {0}")]
+    UnknownError(String),
+
+    #[error(transparent)]
+    BuildError(#[from] metaslang_graph_builder::stack_graph::BuildError),
+}
+
+pub struct Bindings {
+    version: Version,
+    graph_builder_file: graph_builder::File,
+    functions: graph_builder::Functions,
+    stack_graph: StackGraph,
+    cursors: HashMap<GraphHandle, Cursor>,
+}
+
+impl Bindings {
+    pub fn create(version: Version) -> Self {
+        let graph_builder_file = graph_builder::File::from_str(binding_rules::BINDING_RULES_SOURCE)
+            .expect("Bindings stack graph builder parse error");
+        let stack_graph = StackGraph::new();
+        let functions = stack_graph::default_functions();
+        let cursors = HashMap::new();
+
+        Self {
+            version,
+            graph_builder_file,
+            functions,
+            stack_graph,
+            cursors,
+        }
+    }
+
+    pub fn add_file(&mut self, file_path: &str, tree_cursor: Cursor) -> Result<(), BindingsError> {
+        let globals = self.get_globals_for_file(file_path)?;
+        let file = self.stack_graph.get_or_create_file(file_path);
+
+        let mut builder = Builder::new(
+            &self.graph_builder_file,
+            &self.functions,
+            &mut self.stack_graph,
+            file,
+            tree_cursor,
+        );
+        builder.build(&globals, &stack_graph::NoCancellation, |handle, cursor| {
+            self.cursors.insert(handle, cursor.clone());
+        })?;
+
+        Ok(())
+    }
+
+    fn get_globals_for_file(
+        &self,
+        file_path: &str,
+    ) -> Result<graph_builder::Variables<'static>, BindingsError> {
+        let path = PathBuf::from(&file_path).canonicalize()?;
+        let root_path = path.parent().ok_or(BindingsError::UnknownError(
+            "Cannot compute the ROOT_PATH".to_owned(),
+        ))?;
+
+        let mut globals = graph_builder::Variables::new();
```

Comment:
nit:
I assume these globals are used in the bindings file. Is that correct?
I suggest passing them as a strong type to `builder`, and have it set its own variables map if needed, to make sure required properties are set.

Reply:
Additionally, for `ROOT_PATH_VAR`, is this the user project's root path? I wonder if it should be passed to `Bindings::new()` as part of user input to the compilation.

Reply:
Makes sense. These patterns were imported from tree-sitter-stack-graph, but the code will be better if we customize it for our needs.

Reply:
Done. We now have a `Globals` struct which contains the version and the file path. We can add more configuration variables there if necessary, and this formalizes the requirements from our bindings rules files.